pteranodon/server_detector.py

```diff
@@ -0,0 +1,298 @@
+    """
+    Searches serial devices and a server for MAV SDK server instances
+    """
+
+    def __init__(
+        self,
+        server_ip_addr: str = "127.0.0.1",
+        port_range: Tuple[int, int] = (14540, 14550),
+        serial_baud_rate: int = 115200,
+    ):
+        """
+        Searches serial devices and, if a server ip address is provided, that server for open MAV SDK servers.
+
+        :param server_ip_addr: The server IP address to search for MAV SDK searches, None if only serial devices should be searched.
+        :type server_ip_addr: Optional[str]
+        """
+
+        self._event_loop = asyncio.new_event_loop()
+
+        self._server_addr = server_ip_addr
+        self._serial_baud_rate = serial_baud_rate
+
+        # Between socket kind and ports
+        self._available_ports: Dict[int, List[int]] = defaultdict(list)
+
+        self._port_range: Tuple[int, int] = port_range
+
+        if port_range[0] < 0 or port_range[1] < 0 or port_range[0] >= port_range[1]:
+            print("Invalid port range given, reverting to default (14540, 14550)")
+            self._port_range = (14540, 14550)
+
+        self._loop_thread = Thread(
+            name="Detector-Loop-Thread",
+            target=self._run_loop,
+            daemon=True,
+        )
+
+    def _run_loop(self):
+        self._event_loop.run_forever()
+
+    async def _test_port_open(
+        self,
+        port: int,
+        socket_kind: int,
+        timeout: float = 1.0,
+    ) -> Optional[Tuple[int, bool]]:
+        """
+        Attempts to connect to a port with a specified timeout and socket kind to determine if the port is open.
+
+        :param port: The port to check
+        :type port: int
+        :param socket_kind: The type of socket to open (SOCK_STREAM for TCP, SOCK_DGRAM for UDP)
+        :type socket_kind: socket.SocketKind
+        :param timeout: The timeout for the connect call
+        :type timeout: float
+        :return: The port tested and whether the port is open, None if invalid parameters
+        :rtype: Optional[Tuple[int, bool]]
+        """
+
+        # Only supporting TCP and UDP
+        if socket_kind not in [socket.SOCK_STREAM, socket.SOCK_DGRAM]:
+            return None
+
+        with socket.socket(socket.AF_INET, socket_kind) as sock:
+            sock.settimeout(timeout)
+            try:
+                sock.connect((self._server_addr, port))
+            except TimeoutError:
+                return port, False  # Timed out
+            except ConnectionError:
+                return port, False  # Not open
+
+        return port, True
+
+    @staticmethod
+    def _fetch_possible_serial_devs() -> List[str]:
+        """
+        Fetches all the possible serial devices from /sys/class/tty while removing standard devices and virtual consoles.
+
+        :return: The paths of connected serial devices
+        :rtype: List[str]
+        """
+
+        if not os.path.exists("/sys/class/tty"):
+            print(
```

Comment:
Instead of printing lets raise an exception? If printing is wanted we should use a logger...

Can this take a logger optionally? If the logger is passed use that otherwise use logging.wanring or something like that?

Reply:
So, I don't want to raise an exception because it isn't necessarily breaking. Like you can still search the given IP address, so if an exception is raised, then it would be unusable on any system other than Linux.

I think a logger is a good idea, I'll add that.